Conform provenance.materials.uri to URI syntax

zregvart opened the pull request:
<!-- 🎉⛓🎉 Thank you for the PR!!! 🎉⛓🎉 -->

# Changes

According to SLSA Provenance v0.2 the `provenance.materials.uri`[1] needs to be a `ResourceURI` which is defined[2] as a _"Uniform Resource Identifier as specified in RFC 3986"_.

The `uri` field of `provenance.materials` is populated with the image reference which does not contain the scheme and therefore is not a valid URI.

This changes that to prepend the `oci://` prefix to Task and sidecar image references.

[1] https://slsa.dev/provenance/v0.2#materials.uri
[2] https://github.com/in-toto/attestation/blob/main/spec/v0.1.0/field_types.md#ResourceURI

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
SLSA Provenance v0.2 provenance.materials.uri field syntax for image references has the `oci://` prefix.
```

chitrangpatel commented:
/ok-to-test

chitrangpatel commented on test/testdata/slsa/v1/task-output-image.json:
```diff
@@ -44,7 +44,7 @@
 	    {{range .URIDigest}}
 	    {
-		"uri": "{{.URI}}",
+		"uri": "oci://{{.URI}}",
```
May be we should capture this in the URI itself. The reason is that materials can also contain repos of the remote tasks and pipelines which could be on git/oci/artifact hub.

chitrangpatel replied:
I think our example tests inline the spec (instead of fetching them remotely) so they are not throwing errors.

chitrangpatel commented on test/testdata/slsa/v2/task-output-image.json:
```diff
@@ -79,7 +79,7 @@
 	    {{range .URIDigest}}
 	    {
-		"uri": "{{.URI}}",
+		"uri": "oci://{{.URI}}",
```
same comment as above.

chitrangpatel commented on test/testdata/slsa/v1/pipeline-output-image.json:
```diff
@@ -72,7 +72,7 @@
 	    {{range .URIDigest}}
 	    {
-		"uri": "{{.URI}}",
+		"uri": "oci://{{.URI}}",
```
see comment below.

chitrangpatel commented on pkg/artifacts/signable.go:
```diff
@@ -316,7 +316,7 @@ func RetrieveMaterialsFromStructuredResults(ctx context.Context, obj objects.Tek
 		digest := splits[1]
 		mats = append(mats, common.ProvenanceMaterial{
-			URI:    s.URI,
+			URI:    fmt.Sprintf("oci://%s", s.URI),
```
See https://github.com/tektoncd/community/blob/main/teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md for context.

chitrangpatel replied:
Basically its up to the task authors to declare what was fetched.

zregvart commented:
Thanks for the review @chitrangpatel I was overreaching by prefixing the results with `oci://`. I think I've addressed the comments in the latest change I just pushed. Can you have another look?

chuangw6 commented on pkg/chains/formats/slsa/internal/material/material.go:
```diff
@@ -66,7 +66,8 @@ func AddImageIDToMaterials(imageID string, mats *[]common.ProvenanceMaterial) er
 			// no point in partially populating the material
 			// do it if both conditions are valid.
-			m.URI = uriDigest[0]
+			uri := strings.TrimPrefix(uriDigest[0], "docker-pullable://")
+			m.URI = fmt.Sprintf("oci://%s", uri)
```
Thanks @zregvart for fixing this. Can we make the prefix a constant so that we can have a single place to define what the prefixes are? Perhaps we should do it for `git+` as well.

chitrangpatel reviewed (approved):
Thanks!
